rapier/src/pipeline.rs

```diff
@@ -1,19 +1,355 @@
+        ///   surface of the object after traveling through the object interior.
+        ///
+        /// # Panics
+        ///
+        /// This will panic if the `from` point and the `to` point are the same.
+        #[must_use]
+        pub fn ray_cast(&self, from: Vec3, to: Vec3, solid: bool) -> Option<RayCastInfo> {
+            self.ray_cast_internal(from, to, solid, CollisionLayers::default(), None)
+        }
+
+        /// Cast a ray with extra filters
+        ///
+        /// Behaves the same as [`ray_cast()`][Self::ray_cast] but takes extra arguments for
+        /// filtering results:
+        ///
+        /// - `layers`: The [`CollisionLayers`] to considered for collisions, allowing for coarse
+        ///   filtering of collisions.
+        /// - `filter`: A closure taking an [`Entity`] and returning `true` if the entity should be
+        ///   considered for collisions, allowing for fine-grained, per-entity filtering of
+        ///   collisions.
+        ///
+        /// # Panics
+        ///
+        /// This will panic if the `from` point and the `to` point are the same.
+        #[must_use]
+        pub fn ray_cast_with_filter<F>(
+            &self,
+            from: Vec3,
+            to: Vec3,
+            solid: bool,
+            layers: CollisionLayers,
+            filter: F,
+        ) -> Option<RayCastInfo>
+        where
+            F: Fn(Entity) -> bool,
+        {
+            self.ray_cast_internal(from, to, solid, layers, Some(&filter))
+        }
+
+        /// Non-public implementatino of `ray_cast`
+        #[must_use]
+        #[allow(clippy::cast_possible_truncation)]
+        fn ray_cast_internal(
+            &self,
+            from: Vec3,
+            to: Vec3,
+            solid: bool,
+            layers: CollisionLayers,
+            filter: Option<&dyn Fn(Entity) -> bool>,
+        ) -> Option<RayCastInfo> {
+            assert_ne!(from, to, "From and to points must not be the same");
+            let to_from_diff = to - from;
+            let direction = to_from_diff.normalize();
+
+            let ray = Ray::new(from.into_rapier(), direction.into_rapier());
+
+            let result = self.query_pipeline.cast_ray_and_get_normal(
+                &*self.colliders,
+                &ray,
+                to_from_diff.length(),
+                solid,
+                InteractionGroups {
+                    memberships: layers.groups_bits(),
+                    filter: layers.masks_bits(),
+                },
+                // Map filter to one that takes a collider handle and returns a bool
+                filter
+                    .map(|filter| {
+                        move |handle: ColliderHandle| -> bool {
+                            let entity = Entity::from_bits(
+                                self.colliders
+                                    .get(handle)
+                                    .expect("Missing collider")
+                                    .user_data as u64,
```

Comment:
If the collider is not present we can ignore it. Don't panic.

I'm not sure if it can happen at all in practice. But I don't see the point of panicking here ;-)